### ISSUE TITLE
Solving iterative power flow by changing the generator outputs on the fly 

### ISSUE THREAD

**Issue by asavasci** [dss-extensions/OpenDSSDirect.jl#79]
Hi,  I am trying to do a sensitivity analysis study by running the following example script where I first solve the original circuit with base parameters. I want to increase the generator's output by a given rate and then solve the power flow again for the second time with a new generation increment.  After each solve, I collect the voltages to calculate voltage sensitivities. 

It seems the generator output is not being set by `DSS.Generators.kW(DSS.Generators.kW()*rate)`, and I am getting the same voltage values.  

Is there any other thing that I am missing in the code or a better way to do this analysis with the .jl interface?   

`DSS.Text.Command( string("Compile"," ",main_path,"/Master.dss") );`
`DSS.Text.Command( string("set controlmode=OFF") );`

`# Run the first power flow`
`DSS.Solution.Mode(0) # set solution mode as snap=>0, daily=>1`
`DSS.Solution.Solve()`
`println( DSS.Solution.Converged()==true ? "SOLVED" : "NOT SOLVED" )
`
`v0 = DSS.Circuit.AllBusMagPu()  # Record voltage profile`
`println(DSS.Generators.kW())`

`# Run the second power flow by incresing the generator output by a rate`
`DSS.Generators.kW(DSS.Generators.kW()*rate)`
`println(DSS.Generators.kW())`

`DSS.Solution.Mode(0) # set solution mode as snap=>0, daily=>1`
`DSS.Solution.Solve()`
`println( DSS.Solution.Converged()==true ? "SOLVED" : "NOT SOLVED" )`

`v1 = DSS.Circuit.AllBusMagPu() # `

**Comment by PMeira** [comment]
That seems fine and seems to work from a quick test. If you're able to provide an example based on one of the sample circuits, I could look further. 

I imagine there could be some limitation based on how the generators are initially defined.

**Comment by asavasci** [comment]
I am attaching the files here. The generator is defined in the PVSystem.DSS file.
[13Bus_PVSystem_sensitivity.zip](https://github.com/dss-extensions/OpenDSSDirect.jl/files/8985696/13Bus_PVSystem_sensitivity.zip)

**Comment by PMeira** [comment]
From this:
```
New Generator.GEN671  Bus1=671  kV=4.16 kW=1 Pf=1  Model=1 conn=3
```

The initial power is set to 1 kW, that is probably too tiny for this system. Consider that the load in bus 671 is >1000x that:

```
New Load.671 Bus1=671.1.2.3  Phases=3 Conn=Delta Model=1 kV=4.16   kW=1155 kvar=660 
```

If you start with a larger value, you should see changes in the system voltage more easily. E.g. using `kW=200` at first for GEN671:

```
41×2 Matrix{Float64}:
 0.999981  0.999984
 0.999996  1.0
 0.999962  0.999966
 0.999926  0.99993
 0.999973  0.999977
 0.999941  0.999945
 0.99982   0.999829
 0.999895  0.999904
 0.999831  0.999841
 0.955807  0.957227
 0.987482  0.987959
 0.94718   0.948487
 0.931241  0.93262
 0.968108  0.968594
 0.927924  0.929201
 0.958914  0.960339
 0.989523  0.989996
 0.949822  0.951134
 0.948872  0.950727
 0.99203   0.992622
 0.935764  0.937476
 0.928124  0.930749
 0.999683  1.0005
 0.911691  0.914162
 0.928124  0.930749
 0.999683  1.0005
 0.911691  0.914162
 0.948132  0.94944
 0.980177  0.980651
 0.946196  0.947503
 0.978536  0.979009
 0.928124  0.930749
 0.999683  1.0005
 0.911691  0.914162
 0.921583  0.924188
 1.00188   1.0027
 0.909859  0.912324
 0.926316  0.928936
 0.909745  0.91221
 0.907811  0.91027
 0.921084  0.923689
```

**Comment by asavasci** [comment]
Okay, noted!
Thanks.

**Comment by asavasci** [comment]
I would like to put the code into a function and call multiple  when needed. However, I am getting different sensititvity results when I run the code explicitly and within a function. Note: generators rated kW = 200
I am attaching the notebook file.

[main_sensitivity_analysis.ipynb.zip](https://github.com/dss-extensions/OpenDSSDirect.jl/files/8986825/main_sensitivity_analysis.ipynb.zip)

**Comment by PMeira** [comment]
You forgot to reset to the original kW before the first step.

Note that you probably should take more care with your `dV_dP`, but that what to do depends on your goals. You're not updating `v0` after each step of the loop, so it's not a local delta, and the denominator probably should be something like `DSS.Generators.kW() - DSS.Generators.kW()/rate` instead. You could do something like `kW = DSS.Generators.kW()` to avoid the API call and clean up the code.

**Comment by asavasci** [comment]
I didn't quite understand why I should use `DSS.Generators.kW() - DSS.Generators.kW()/rate` since I wanted to increase the output of the generator. Also, I wanted to observe the effect of increment on the base solution. Is there any specific reason for updating v0 after each step of the loop?

**Comment by PMeira** [comment]
> Does DSS.Generators.kW(DSS.Generators.kW()*rate) only set the value but it does not modify the generator's output unless the DSS.Solution.Solve() is run?

Ah, I see what misled you. That actually is somewhat of a bug that was only fixed in the latest OpenDSS release (9.4 IIRC). We port the change in https://github.com/dss-extensions/OpenDSSDirect.jl/pull/73

OpenDSS exposes the current running state of the element and system properties, but some values are propagated only after the solution. The current API mixes the plain object properties and running state in some properties/functions with same name, which is confusing and sometimes misleading. Since we have to match the official OpenDSS behavior for the current APIs, we cannot change too much here. We'll probably expose an alternative API in the medium term.

> Also, I wanted to observe the effect of increment on the base solution.

If you want to always compare to the base solution, I think you wanted something like "(current v - initial v) / (current kW - initial kW)". In this case, v0="initial v" would be constant, but so should be initial kW -- currently it's a function of the initial kW and all previous rate values. This...
```
DSS.Generators.kW(DSS.Generators.kW()*rate) 
```
...is basically setting the kW to (initial kW) * rate[1] * rate[2] * rate[3] ... * rate[n] at the end of the loop.

Anyway, I would avoid calling API functions if not necessary. Each call to the DSS API will have an overhead (Julia to/from Pascal layer, various state checks in the DSS engine, etc.), so it's a good habit to structure the code to minimize unnecessary calls, otherwise you may have performance issues if you want to run a large scale system someday.

**Comment by PMeira** [comment]
> It would be nice if there is a cheat sheet for using API functions within more examples so that a user can get some idea of good coding habits for run-time performance.

I agree, we'll focus on that for the new API (which is lighter too). I'm working on that this week again, finally. The original API will be kept for compatibility.

(Closing the issue since I think we exhausted the original topic.)
